Don't sync assignment events for gates

jmileham opened the pull request:
Fixes #43 

Because feature gates aren't needed for A/B test analysis and [TestTrack server will now discard the assignment event](https://github.com/Betterment/test_track/pull/69
) in order to allow visitors to dynamically reassign upon gate weighting changes, we just deterministically assign locally each time instead and skip notification of the event.

This code change is pretty far from the notification code, but leans on the fact that all notification codepaths rely on visitor.unsynced_assignments to decide whether to send assignment notifications, which in turn calls the `unsynced?` method on assignment, which was previously hard-coded to return true - basically if the rails client made an assignment, it was considered worthy of syncing to the server before. Now we will consider feature gate assignments to be synced, so they will not trigger a notification.

This should make pages and jobs that only have feature gates (and not experiment assignments) less resource intensive than ever!

jmileham commented on app/models/test_track/assignment.rb:
```diff
@@ -16,7 +16,11 @@ def variant
+  end
+
+  def feature_gate?
+    split_name.end_with?('_enabled')
```
The wire protocol TestTrack uses for split registries is too concise to add an explicit field describing gate-iness without changing it completely. Currently it's like so:

```JSON
{
  "split_name": { "variant1": 50, "variant2": 50 }
}
```

So I'm planning on making `_enabled` a hard requirement of feature gate naming as we advance from here so that the client can reliably infer gate-iness this way.

jmileham commented:
Note that this affects both metrics collection assignment notifications and TT server notifications. @asullivan210 I had been misremembering the way the rails client worked. I thought it sent an experience event too, but realized now that TT doesn't consider itself responsible for experience events, only assignment events.

So this change will make feature gate events go dark unless you instrumented your app code per-feature. At the same time, continuing to send assignment events for feature gates as we were seems like a lie, because they'd become too chatty. Maybe we should think about adding a split experienced event *instead*? Or maybe we can just use generalized app analytics to track traffic on the page we know to be gated if we want to know how much traffic its getting?

jmileham commented:
/domain joejansen given this comment: https://github.com/Betterment/test_track/pull/69#comment-181493116 you’ll definitely have a horse in this race then.

This PR doesn’t block the TT server PR so we can mess with this til it does satisfying things.

jmileham commented:
/domain @joejansen that is ^^

samandmoore commented on app/models/test_track/assignment.rb:
```diff
@@ -16,7 +16,11 @@ def variant
   end
 
   def unsynced?
```
i think we should rename this to `should_sync?`. since some `Assignment`s are no longer meant to be synced, it seems to me that this name is misleading as is. i.e. feature gates are intentionally "unsynced" so changing the naming here will provide a clearer indication of that.

samandmoore commented:
<<domainlgtm

i reasoned through this for a while, and this seems to be the least invasive and simplest way to change the behavior. 

just change that method name 👍

jmileham commented:
Cool so the last bit is how we want to ensure devs get a sense for how many people are experiencing which fork of a gate. Sam and I were thinking maybe there’s another event we want to send to the event bus indicating that the client selected variant X of feature gate Y. @joejansen thoughts?

joejansen commented:
i think there is value in having some analytics event sent to segment or whatever analytics client the test track clients configure. when we ship something that relies on a feature gate, an analytics event can help us confirm that things are working as we intend and in the proportion or absolute volume we'd expect. in many cases we use feature gates in model code that will be triggered by a delayed job or a system transaction where there is no other analytics event that can serve as a good proxy. further, even when there is a proxy event, we may be relying on some filtering logic that only exposes a portion of users to the gate based on state the proxy event is unaware of. 

i understand the event name `SplitAssigned` no longer fits as well given that we are no longer recording an assignment in the server application. i'd be in favor of implementing a new analytics event like `SplitExperienced` or something even more specific to this use case like `FeatureFlagExperienced` or `FeatureFlagExercised`.

joejansen commented:
<<domain tafn

jmileham commented:
bump. Total change of approach here.

Now for feature gates we skip TT assignment events but continue to send an analytics event. The built-in mixpanel client now sends a `FeatureGateExperienced` event instead of a `SplitAssigned` event. Consuming apps will have to modify their custom analytics clients to differentiate event types like this, but if they fail to, they'll just get noisey assignments for longer. Not the end of the world.

 @samandmoore you'll probably want to give this another look.

jmileham commented on app/models/test_track/analytics/mixpanel_client.rb:
```diff
@@ -11,6 +11,14 @@ def mixpanel
+      else
+        'SplitAssigned'
+      end
+    end
```
I could stuff this function into assignment as a method, e.g. `analytics_event_name`. It'd be your option whether you choose use it in your custom analytics plugin or use something else but you wouldn't have to duplicate the biz logic then.

jmileham commented:
bump - I revoked your LGs. Sorry. But I changed too much not to get another look.

Ok, so this just became a significantly breaking change to the custom analytics client API contract (like releasing client v4.0-level breaking change), but I think that's OK because you'd want to get on board with the event name changes for feature gates anyway, and not having a forcing function to reassess when upgrading the TT client would probably result in confusion.

After talking with @samandmoore offline I changed the `AnalyticsClient#track_assignment(visitor_id, assignment)` contract to `AnalyticsClient#track(analytics_event)`. `TestTrack::AnalyticsEvent` provides the business logic to alternate between feature gate/experiment event names and provides a ready-to-use property payload so that the analytics client merely needs to figure out how to instantiate the tracking lib and call it with the right args instead of do a bunch of data munging and hash construction per plugin.

I also updated the readme to match and dropped upgrading to 4.0 instructions.

I don't actually intend to release this as 4.0 until we also make multi-identity sessions a thing in order to allow lookups of variants of other people from model code.

jmileham commented:
nanda?

samandmoore commented on spec/models/test_track/analytics_event_spec.rb:
```diff
@@ -0,0 +1,50 @@
+      expect(subject.name).to eq "SplitAssigned"
+    end
+  end
+end
```
Do we care about documenting the public interface of this class? For example, that it exposes assignment and visitor_id as attrs. 

Since we don’t have api docs, the only real proof that we won’t rip out some public readers is a test case. 

It’s sorta documented through the other test that uses this object as a double, but that feels pretty indirect.

jmileham replied:
cool, i reorged the tests around typical `describe "#method_name"` style and hit all the public interface.

samandmoore commented on spec/models/test_track/analytics/safe_wrapper_spec.rb:
```diff
@@ -1,22 +1,22 @@
     it 'calls underlying' do
-      expect(subject.track_assignment(123, foo: "bar")).to eq true
-      expect(underlying).to have_received(:track_assignment).with(123, foo: "bar")
+      expect(subject.track("my event")).to eq true
```
We don’t care that this is no longer a true representation of the contract do we? I think it’s fine.

jmileham replied:
do you mean you'd prefer I use a mock that looks more like an event?

samandmoore commented on spec/models/test_track/assignment_spec.rb:
```diff
@@ -48,4 +48,18 @@
+        expect(subject.unsynced?).to eq true
+      end
+    end
+  end
```
Keeping this test around for clarity?

jmileham replied:
yeah, now that we've decided that `unsynced?` means something else, it felt good to leave this test to lock it down. I also discovered that the fake assignment does leverage this to return false, so everything was getting gnarly.

samandmoore commented:
<<domainlgtm

jmileham commented:
bump @joejansen this cool w/you?
